File: pageindicatorview/src/main/java/com/rd/draw/drawer/Drawer.java
package com.rd.draw.drawer;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.support.annotation.NonNull;
import com.rd.animation.data.Value;
import com.rd.draw.data.Indicator;
import com.rd.draw.drawer.type.*;

public class Drawer {

    private BasicDrawer basicDrawer;
    private ColorDrawer colorDrawer;
    private ScaleDrawer scaleDrawer;
    private WormDrawer wormDrawer;
    private SlideDrawer slideDrawer;
    private FillDrawer fillDrawer;
    private ThinWormDrawer thinWormDrawer;
    private DropDrawer dropDrawer;
    private SwapDrawer swapDrawer;
    private ScaleDownDrawer scaleDownDrawer;

    private int position;
    private int coordinateX;
    private int coordinateY;

    public Drawer(@NonNull Indicator indicator) {
        Paint paint = new Paint();
        paint.setStyle(Paint.Style.FILL);
        paint.setAntiAlias(true);

        basicDrawer = new BasicDrawer(paint, indicator);
        colorDrawer = new ColorDrawer(paint, indicator);
        scaleDrawer = new ScaleDrawer(paint, indicator);
        wormDrawer = new WormDrawer(paint, indicator);
        slideDrawer = new SlideDrawer(paint, indicator);
        fillDrawer = new FillDrawer(paint, indicator);
        thinWormDrawer = new ThinWormDrawer(paint, indicator);
        dropDrawer = new DropDrawer(paint, indicator);
        swapDrawer = new SwapDrawer(paint, indicator);
        scaleDownDrawer = new ScaleDownDrawer(paint, indicator);
    }

    public void setup(int position, int coordinateX, int coordinateY) {
        this.position = position;
        this.coordinateX = coordinateX;
        this.coordinateY = coordinateY;
    }

    public void drawBasic(@NonNull Canvas canvas, boolean isSelectedItem) {
        if (colorDrawer != null) {
            basicDrawer.draw(canvas, position, isSelectedItem, coordinateX, coordinateY);
        }
    }

    public void drawColor(@NonNull Canvas canvas, @NonNull Value value) {
        if (colorDrawer != null) {
            colorDrawer.draw(canvas, value, position, coordinateX, coordinateY);
        }
    }

    public void drawScale(@NonNull Canvas canvas, @NonNull Value value) {
        if (scaleDrawer != null) {
            scaleDrawer.draw(canvas, value, position, coordinateX, coordinateY);
        }
    }

    public void drawWorm(@NonNull Canvas canvas, @NonNull Value value) {
        if (wormDrawer != null) {
            wormDrawer.draw(canvas, value, coordinateX, coordinateY, position);
        }
    }

    public void drawSlide(@NonNull Canvas canvas, @NonNull Value value) {
        if (slideDrawer != null) {
            slideDrawer.draw(canvas, value, coordinateX, coordinateY);
        }
    }

    public void drawFill(@NonNull Canvas canvas, @NonNull Value value) {
        if (fillDrawer != null) {
            fillDrawer.draw(canvas, value, position, coordinateX, coordinateY);
        }
    }

    public void drawThinWorm(@NonNull Canvas canvas, @NonNull Value value) {
        if (thinWormDrawer != null) {
            thinWormDrawer.draw(canvas, value, coordinateX, coordinateY);
        }
    }

    public void drawDrop(@NonNull Canvas canvas, @NonNull Value value) {
        if (dropDrawer != null) {
            dropDrawer.draw(canvas, value, coordinateX, coordinateY);
        }
    }

    public void drawSwap(@NonNull Canvas canvas, @NonNull Value value) {
        if (swapDrawer != null) {
            swapDrawer.draw(canvas, value, position, coordinateX, coordinateY);
        }
    }

    public void drawScaleDown(@NonNull Canvas canvas, @NonNull Value value) {
        if (scaleDownDrawer != null) {
            scaleDownDrawer.draw(canvas, value, position, coordinateX, coordinateY);
        }
    }
}

File: pageindicatorview/src/main/java/com/rd/draw/drawer/Drawer.kt
package com.rd.draw.drawer

import android.graphics.Canvas
import android.graphics.Paint
import com.rd.animation.data.Value
import com.rd.draw.data.Indicator
import com.rd.draw.drawer.type.*

class Drawer(indicator: Indicator) {

	private val paint = Paint()
	private val basicDrawer = BasicDrawer(paint, indicator)
	private val colorDrawer = ColorDrawer(paint, indicator)
	private val scaleDrawer = ScaleDrawer(paint, indicator)
	private val wormDrawer = WormDrawer(paint, indicator)
	private val slideDrawer = SlideDrawer(paint, indicator)
	private val fillDrawer = FillDrawer(paint, indicator)
	private val thinWormDrawer = ThinWormDrawer(paint, indicator)
	private val dropDrawer = DropDrawer(paint, indicator)
	private val swapDrawer = SwapDrawer(paint, indicator)
	private val scaleDownDrawer = ScaleDownDrawer(paint, indicator)

	private var position: Int = 0
	private var coordinateX: Int = 0
	private var coordinateY: Int = 0

	init {
		paint.style = Paint.Style.FILL
		paint.isAntiAlias = true
	}

	fun setup(position: Int, coordinateX: Int, coordinateY: Int) {
		this.position = position
		this.coordinateX = coordinateX
		this.coordinateY = coordinateY
	}

	fun drawBasic(canvas: Canvas, isSelectedItem: Boolean) {
		basicDrawer.draw(canvas, position, isSelectedItem, coordinateX, coordinateY)
	}

	fun drawColor(canvas: Canvas, value: Value) {
		colorDrawer.draw(canvas, value, position, coordinateX, coordinateY)
	}

	fun drawScale(canvas: Canvas, value: Value) {
		scaleDrawer.draw(canvas, value, position, coordinateX, coordinateY)
	}

	fun drawWorm(canvas: Canvas, value: Value) {
		wormDrawer.draw(canvas, value, coordinateX, coordinateY)
	}

	fun drawSlide(canvas: Canvas, value: Value) {
		slideDrawer.draw(canvas, value, coordinateX, coordinateY)
	}

	fun drawFill(canvas: Canvas, value: Value) {
		fillDrawer.draw(canvas, value, position, coordinateX, coordinateY)
	}

	fun drawThinWorm(canvas: Canvas, value: Value) {
		thinWormDrawer.draw(canvas, value, coordinateX, coordinateY)
	}

	fun drawDrop(canvas: Canvas, value: Value) {
		dropDrawer.draw(canvas, value, coordinateX, coordinateY)
	}

	fun drawSwap(canvas: Canvas, value: Value) {
		swapDrawer.draw(canvas, value, position, coordinateX, coordinateY)
	}

	fun drawScaleDown(canvas: Canvas, value: Value) {
		scaleDownDrawer.draw(canvas, value, position, coordinateX, coordinateY)
	}
}
